pkg/k8s/util.go

```diff
@@ -37,15 +38,72 @@ func GetHostnamesForService(svc service.MeshService, localNamespace bool) []stri
 	return hostnames
 }
 
-// GetServiceFromHostname returns the service name from its hostname
-func GetServiceFromHostname(host string) string {
-	// The service name is the first string in the host name for a service.
-	// Ex. service.namespace, service.namespace.cluster.local
-	service := strings.Split(host, ".")[0]
+// splitHostName takes a k8s FQDN (i.e. host) and retrieves the service name
+// as well as the subdomain (may be empty)
+func splitHostName(host string) (service string, subdomain string) {
+	serviceComponents := strings.Split(host, ".")
+
+	// The service name is usually the first string in the host name for a service.
+	// Ex. service.namespace, service.namespace.svc.cluster.local
+	// However, if there's a subdomain, we the service name is the second string.
+	// Ex. mysql-0.service.namespace, mysql-0.service.namespace.svc.cluster.local, mysql-0.service.namespace.svc.cluster.local
+	switch l := len(serviceComponents); {
+	case l == 1:
+		// e.g. service
+		service = serviceComponents[0]
+		subdomain = ""
+	case l == 2:
+		// e.g. service.namespace, mysql-0.service
+		// NOTE: There's no way to tell which component is the service or not, so we're going to keep the default behavior
+		// Thus, we have a requirement that users who want to reference statefulsets must include the namespace in the host name
+		service = serviceComponents[0]
+		subdomain = ""
+	case l == 3:
+		if strings.Contains(serviceComponents[l-1], "svc") {
```

Comment:
The contains is there so that if there's a port at the end (e.g. `mysql-0.service.svc:9090`), the condition still evaluates to true. I see what you mean though; I'll write some more logic to account for that use-case. 

As to your other question, I don't even think Kubernetes supports a namespace called `svc` due to how it resolves DNS.

Reply:
Following the example here to spin up a basic StatefulSet in a namespace called `svc` in a kind cluster: https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/

`nslookup` successfully resolves all of the following:
```
web-0.nginx
web-0.nginx.svc
web-0.nginx.svc.svc
web-0.nginx.svc.svc.cluster.local
```

From another namespace also with a service named `nginx`, I see `nslookup nginx.svc` and `nslookup web-0.nginx.svc` consistently resolve to only the `nginx` in the `svc` namespace instead of the `nginx` in the same namespace. I don't think I've seen that exact behavior documented so I'm not sure exactly what we should expect there, but it's probably worth making sure the behavior here matches that.

Reply:
Wow, I wouldn't have expected that; thanks for digging in. This does become more difficult though since there's no way to distinguish between `web-0.nginx.svc` where svc is the namespace and `web-0.nginx.svc` where web-0 is the actual service name and svc is the TLD.

Reply:
I've been thinking about this some more, and I still can't see a away to distinguish between these two cases. Maybe we just have to make it a known limitation that namespaces named `svc` may result in unexpected behavior?

Reply:
@nojnhuh I ended up taking Sean's approach and used a Controller to make this split more accurae